satpy/data_download.py

```diff
@@ -0,0 +1,356 @@
+                           "file '{}' does not exist.".format(local_file))
+    return local_file
+
+
+def retrieve(cache_key, pooch_kwargs=None):
+    """Download and cache the file associated with the provided ``cache_key``.
+
+    Cache location is controlled by the config ``data_dir`` key. See
+    :ref:`data_dir_setting` for more information.
+
+    Args:
+        cache_key (str): Cache key returned by
+            :func:`~satpy.data_download.register_file`.
+        pooch_kwargs (dict or None): Extra keyword arguments to pass to
+            :meth:`pooch.Pooch.fetch`.
+
+    Returns:
+        Local path of the cached file.
+
+
+    """
+    pooch_kwargs = pooch_kwargs or {}
+
+    path = satpy.config.get('data_dir')
+    if not satpy.config.get('download_aux'):
+        return _retrieve_offline(path, cache_key)
+    # reuse data directory as the default URL where files can be downloaded from
+    pooch_obj = pooch.create(path, path, registry=_FILE_REGISTRY,
+                             urls=_FILE_URLS)
+    return pooch_obj.fetch(cache_key, **pooch_kwargs)
+
+
+def retrieve_all(readers=None, writers=None, composite_sensors=None,
+                 pooch_kwargs=None):
+    """Find cache-able data files for Satpy and download them.
+
+    The typical use case for this function is to download all ancillary files
+    before going to an environment/system that does not have internet access.
+
+    Args:
+        readers (list or None): Limit searching to these readers. If not
+            specified or ``None`` then all readers are searched. If an
+            empty list then no readers are searched.
+        writers (list or None): Limit searching to these writers. If not
+            specified or ``None`` then all writers are searched. If an
+            empty list then no writers are searched.
+        composite_sensors (list or None): Limit searching to composite
+            configuration files for these sensors. If ``None`` then all sensor
+            configs will be searched. If an empty list then no composites
+            will be searched.
+        pooch_kwargs (dict): Additional keyword arguments to pass to pooch
+            ``fetch``.
+
+    """
+    if pooch_kwargs is None:
+        pooch_kwargs = {}
+    if not satpy.config.get('download_aux'):
+        raise RuntimeError("Satpy 'download_aux' setting is False so no files "
+                           "will be downloaded.")
+
+    find_registerable_files(readers=readers,
+                            writers=writers,
+                            composite_sensors=composite_sensors)
+    path = satpy.config.get('data_dir')
+    pooch_obj = pooch.create(path, path, registry=_FILE_REGISTRY,
+                             urls=_FILE_URLS)
+    for fname in _FILE_REGISTRY:
+        logger.info("Downloading extra data file '%s'...", fname)
+        pooch_obj.fetch(fname, **pooch_kwargs)
+    logger.info("Done downloading all extra files.")
```

Comment:
The abstraction levels here aren't really consistent...

Reply:
Not sure what you mean? What do you suggest I do differently? Put pooch creation and for loop in a separate function?

Reply:
yes, 146-151 can be a separate method.